esp-hal/src/lcd_cam/cam.rs

```diff
@@ -1,5 +1,671 @@
-use crate::{peripheral::PeripheralRef, peripherals::LCD_CAM};
+//! # Camera - Master or Slave Mode
+//!
+//! ## Overview
+//! The LCD_CAM peripheral supports receiving 8/16 bit DVP signals in either
+//! master or slave mode. In master mode, the peripheral provides the master
+//! clock to drive the camera, in slave mode it does not. This is configured
+//! with the `with_master_clock` method on the camera driver. The driver (due to
+//! the peripheral) mandates DMA (Direct Memory Access) for efficient data
+//! transfer.
+//!
+//! ## Examples
+//! Following code shows how to receive some bytes from an 8 bit DVP stream in
+//! master mode.
+//!
+//! ```no_run
+//! let mclk_pin = io.pins.gpio15;
+//! let vsync_pin = io.pins.gpio6;
+//! let href_pin = io.pins.gpio7;
+//! let pclk_pin = io.pins.gpio13;
+//! let data_pins = RxEightBits::new(
+//!     io.pins.gpio11,
+//!     io.pins.gpio9,
+//!     io.pins.gpio8,
+//!     io.pins.gpio10,
+//!     io.pins.gpio12,
+//!     io.pins.gpio18,
+//!     io.pins.gpio17,
+//!     io.pins.gpio16,
+//! );
+//!
+//! let lcd_cam = LcdCam::new(peripherals.LCD_CAM);
+//! let mut camera = Camera::new(lcd_cam.cam, channel.rx, data_pins, 20u32.MHz(), &clocks)
+//!     .with_master_clock(mclk_pin) // Remove this for slave mode.
+//!     .with_ctrl_pins(vsync_pin, href_pin, pclk_pin);
+//! ```
+
+use core::mem::size_of;
+
+use embedded_dma::WriteBuffer;
+use fugit::HertzU32;
+
+use crate::{
+    clock::Clocks,
+    dma::{
+        ChannelRx,
+        ChannelTypes,
+        DmaError,
+        DmaPeripheral,
+        DmaTransfer,
+        LcdCamPeripheral,
+        RegisterAccess,
+        Rx,
+        RxChannel,
+        RxPrivate,
+    },
+    gpio::{InputPin, InputSignal, OutputPin, OutputSignal},
+    i2s::Error,
+    lcd_cam::{cam::private::RxPins, private::calculate_clkm, BitOrder, ByteOrder},
+    peripheral::{Peripheral, PeripheralRef},
+    peripherals::LCD_CAM,
+};
+
+/// Generation of GDMA SUC EOF
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum EofMode {
+    /// Generate GDMA SUC EOF by data byte length
+    ByteLen,
+    /// Generate GDMA SUC EOF by the vsync signal
+    VsyncSignal,
+}
+
+/// Vsync Filter Threshold
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum VsyncFilterThreshold {
+    One,
+    Two,
+    Three,
+    Four,
+    Five,
+    Six,
+    Seven,
+    Eight,
+}
 
 pub struct Cam<'d> {
-    pub(crate) _lcd_cam: PeripheralRef<'d, LCD_CAM>,
+    pub(crate) lcd_cam: PeripheralRef<'d, LCD_CAM>,
+}
+
+pub struct Camera<'d, RX, P> {
+    lcd_cam: PeripheralRef<'d, LCD_CAM>,
+    rx_channel: RX,
+    _pins: P,
```

Comment:
There's some type checking of `P` in the read_dma call itself, where I'm basically checking for 8 bit vs 16 bit.

I'm happy to erase/simplify the generics but don't know how to handle bus width without this.

Reply:
Hmm, if the only check is to figure out the bit width of the transfer, then could we just store the bit length inside the camera struct as an usize? `type Word` could become `const BIT_WIDTH: usize` and stored directly from `new`?

Feel free to do it a different way, but I think in general we can afford to a small runtime check here (and if the compiler is smart it will see that for any given creation of `Camera`, it will always be `P::BIT_WIDTH` and optimize accordingly)

Reply:
A C style enum with each bit width value would probably work quite nicely too.

Reply:
Sounds good to me, I'll update this too. Will make it much easier to name the type.

Reply:
Actually what do you think of the way it looks now?

I was trying to move the `P` parameter to just the c'tor but the compiler basically asking me to bring back the `RxPins` trait which I was quite happy to get rid of.

Reply:
Hmm don't we need some kind of constraint on `P` anyway though? Unless I'm missing something, what's stopping me passing in anything into the pins field? I.e `Camera<'d, DMA, ()>` would type check as it stands? I suppose `read_dma` wouldn't be callable because its only implemented on certain Rx bit structs, but that would be quite confusing to debug.

My main point for removing the pin generics is that does the user creating the camera care if the camera is 8bit or 16bit _after_ its created? If the answer is no, and I think the answer is no, then adding that type information is just noise.

What do you think?

Reply:
I couldn't think of a good reason for users to care about the Camera width at compile time so I've removed it.

Yeah you're right about the constraint on `P`, my newness to Rust is showing :sweat_smile: .